File: src/main/java/mutation/testing/examples/order/processor/OrderProcessor.java
package mutation.testing.examples.order.processor;

import mutation.testing.examples.order.processor.article.OrderArticle;
import mutation.testing.examples.order.processor.billing.BillingSystem;
import mutation.testing.examples.order.processor.ordernumber.OrderNumberCreator;

public class OrderProcessor {

    private BillingSystem billingSystem;
    private OrderNumberCreator orderNumberCreator;

    public OrderProcessor(BillingSystem billingSystem, OrderNumberCreator orderNumberCreator) {
        this.billingSystem = billingSystem;
        this.orderNumberCreator = orderNumberCreator;
    }

    public String processOrder(OrderArticle article, int quantity) {
        String name = article.getName();
        billingSystem.bill(makeUpper(name), article.getPrice(), quantity);
        return orderNumberCreator.createOrderNumber();
    }

    private String makeUpper(String name) {
        return name.toUpperCase();
    }

}

File: ruby/lib/order_processor.rb
class OrderProcessor
  def initialize(billing_system, order_number_provider)
    @billing_system = billing_system
    @order_number_provider = order_number_provider
  end

  def process_order(article, quantity)
    @billing_system.bill(article.name, article.price, quantity)
    @order_number_provider.create_order_number()
  end
end

class OrderArticle
  attr_reader :name, :price
  def initialize(name, price)
    @name = name
    @price = price
  end
end
